From 0c61e30f07110da37922a8e41b1e5959fd41a248 Mon Sep 17 00:00:00 2001
From: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Date: Mon, 22 Dec 2014 16:08:33 -0500
Subject: [input] alps: v7 - ignore new packets

Message-id: <1419264516-15459-2-git-send-email-benjamin.tissoires@redhat.com>
Patchwork-id: 102382
O-Subject: [PATCH RHEL7.1 1/4] Input: alps - v7: ignore new packets
Bugzilla: 1107819
RH-Acked-by: Hans de Goede <hdegoede@redhat.com>
RH-Acked-by: Aristeu S. Rozanski F. <arozansk@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1107819
Upstream: will be shipped in v3.20
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8420761
Tested by me on a Toshiba Tecra W50-A with a v7 Alps touchpad

commit 8b23811535d2e1dd6abbe4ce6ea1edfd50ce72de
(in the input maintainer's tree:
https://git.kernel.org/cgit/linux/kernel/git/dtor/input.git)
Author: Hans de Goede <hdegoede@redhat.com>
Date:   Thu Dec 18 09:52:59 2014 -0800

    Input: alps - v7: ignore new packets

    NEW packets are send to indicate a discontinuity in the finger coordinate
    reporting. Specifically a finger may have moved from slot 0 to 1 or vice
    versa.  INPUT_MT_TRACK takes care of this for us.

    NEW packets have 3 problems:
    1) They do not contain middle / right button info (on non clickpads)
       this can be worked around by preserving the old button state
    2) They do not contain an accurate fingercount, and they are
       typically send when the number of fingers changes. We cannot use
       the old finger count as that may mismatch with the amount of
       touch coordinates we've available in the NEW packet
    3) Their x data for the second touch is inaccurate leading to
       a possible jump of the x coordinate by 16 units when the first
       non NEW packet comes in

    Since problems 2 & 3 cannot be worked around, just ignore them.

    BugLink: https://bugs.freedesktop.org/show_bug.cgi?id=86338
    Cc: stable@vger.kernel.org # 3.17
    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Tested-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
    Signed-off-by: Dmitry Torokhov <dmitry.torokhov@gmail.com>

Signed-off-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/input/mouse/alps.c b/drivers/input/mouse/alps.c
index a59a1a6..b3d178d 100644
--- a/drivers/input/mouse/alps.c
+++ b/drivers/input/mouse/alps.c
@@ -938,18 +938,36 @@ static int alps_decode_packet_v7(struct alps_fields *f,
   return 0;
  if (pkt_id == V7_PACKET_ID_UNKNOWN)
   return -1;
+ /*
+  * NEW packets are send to indicate a discontinuity in the finger
+  * coordinate reporting. Specifically a finger may have moved from
+  * slot 0 to 1 or vice versa. INPUT_MT_TRACK takes care of this for
+  * us.
+  *
+  * NEW packets have 3 problems:
+  * 1) They do not contain middle / right button info (on non clickpads)
+  *    this can be worked around by preserving the old button state
+  * 2) They do not contain an accurate fingercount, and they are
+  *    typically send when the number of fingers changes. We cannot use
+  *    the old finger count as that may mismatch with the amount of
+  *    touch coordinates we've available in the NEW packet
+  * 3) Their x data for the second touch is inaccurate leading to
+  *    a possible jump of the x coordinate by 16 units when the first
+  *    non NEW packet comes in
+  * Since problems 2 & 3 cannot be worked around, just ignore them.
+  */
+ if (pkt_id == V7_PACKET_ID_NEW)
+  return 1;
 
  alps_get_finger_coordinate_v7(f->mt, p, pkt_id);
 
- if (pkt_id == V7_PACKET_ID_TWO || pkt_id == V7_PACKET_ID_MULTI) {
-  f->left = (p[0] & 0x80) >> 7;
-  f->right = (p[0] & 0x20) >> 5;
-  f->middle = (p[0] & 0x10) >> 4;
- }
+ f->left = (p[0] & 0x80) >> 7;
+ f->right = (p[0] & 0x20) >> 5;
+ f->middle = (p[0] & 0x10) >> 4;
 
  if (pkt_id == V7_PACKET_ID_TWO)
   f->fingers = alps_get_mt_count(f->mt);
- else if (pkt_id == V7_PACKET_ID_MULTI)
+ else /* pkt_id == V7_PACKET_ID_MULTI */
   f->fingers = 3 + (p[5] & 0x03);
 
  return 0;
-- 
1.7.1